applications/StructuralMechanicsApplication/custom_response_functions/adjoint_structural_solver.py

```diff
@@ -0,0 +1,156 @@
+            else:
+                raise Exception("Invalid DOMAIN_SIZE: " + str(domain_size))
+        else:
+            raise Exception("invalid response_type: " + self.response_function_settings["response_type"].GetString())
+
+        super(AdjointStructuralSolver, self).Initialize()
+
+        self.print_on_rank_zero("::[AdjointMechanicalSolver]:: ", "Finished initialization.")
+
+    def Solve(self):
+        if self.response_function_settings["response_type"].GetString() == "adjoint_strain_energy":
+            self._SolveSpecialStrainEnergy()
+        else:
+            if self.settings["clear_storage"].GetBool():
+                self.Clear()
+            mechanical_solution_strategy = self.get_mechanical_solution_strategy()
+            mechanical_solution_strategy.Solve()
+
+    def _SolveSpecialStrainEnergy(self):
+
+        self.response_function.Initialize()
+
+        for node in self.main_model_part.Nodes:
+            disp_x = node.GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_X,0)
+            disp_y = node.GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Y,0)
+            disp_z = node.GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Z,0)
+            node.SetSolutionStepValue(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT_X,0,disp_x * 0.5)
+            node.SetSolutionStepValue(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT_Y,0,disp_y * 0.5)
+            node.SetSolutionStepValue(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT_Z,0,disp_z * 0.5)
+            if self.settings["rotation_dofs"].GetBool():
+                rot_x = node.GetSolutionStepValue(KratosMultiphysics.ROTATION_X,0)
+                rot_y = node.GetSolutionStepValue(KratosMultiphysics.ROTATION_Y,0)
+                rot_z = node.GetSolutionStepValue(KratosMultiphysics.ROTATION_Z,0)
+                node.SetSolutionStepValue(StructuralMechanicsApplication.ADJOINT_ROTATION_X,0,rot_x * 0.5)
+                node.SetSolutionStepValue(StructuralMechanicsApplication.ADJOINT_ROTATION_Y,0,rot_y * 0.5)
+                node.SetSolutionStepValue(StructuralMechanicsApplication.ADJOINT_ROTATION_Z,0,rot_z * 0.5)
+
+        self.response_function.FinalizeSolutionStep()
+
+    def _create_mechanical_solution_strategy(self):
+        computing_model_part = self.GetComputingModelPart()
+        mechanical_scheme = self.get_solution_scheme()
+        linear_solver = self.get_linear_solver()
+        builder_and_solver = self.get_builder_and_solver()
+        return KratosMultiphysics.ResidualBasedLinearStrategy(computing_model_part,
+                                                              mechanical_scheme,
+                                                              linear_solver,
+                                                              builder_and_solver,
+                                                              False,
```

Comment:
this is not very verbose, could you please add variables with the flag names and pass those?